### PR TITLE
Correct off-by-one problem in EKF IMU ringbuffer

### LINKED ISSUE #25316: EKF_Buffer: there is a bug in ekf_imu_buffer::push_youngest_element(const void *element)
Hi, in ekf_imu_buffer::push_youngest_element(const void *element):
https://github.com/ArduPilot/ardupilot/blob/5b34a9bf1370186c1aafe7f9f35dc77235401a9f/libraries/AP_NavEKF/EKF_Buffer.cpp#L159
```
void ekf_imu_buffer::push_youngest_element(const void *element)
{
    if (!buffer) {
        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
        return;
    }
    // push youngest to the buffer
    _youngest = (_youngest+1) % _size;
    memcpy(get_offset(_youngest), element, elsize);
    // set oldest data index
    _oldest = (_youngest+1) % _size;
    if (_oldest == 0) {                    // <=============BUG!!!
        _filled = true;
    }
}
```
, I think 
```
    if (_oldest == 0) {
        _filled = true;
    }
```
should be modified as:
```
    if (_oldest == 1 || _size == 1) {
        _filled = true;
    }
```

Why?

because when push, the first element is pushed into pos 1 NOT pos 2:
```
_youngest = (_youngest+1) % _size;  // here _youngest is initialized as 0, so the first elem pos is 1 not 0
```

so when _oldest is 0, there is N-1 elem not N elem in buffer:
` _oldest = (_youngest+1) % _size;` 
In fact, the buffer is not full.

I use a pic to describe this bug:
image

### PR DESCRIPTION
**Pull request by peterbarker**
This is an alternate solution to a bug @luweiagi found in the EKF buffer code.

His PR is here: https://github.com/ArduPilot/ardupilot/issues/25316

My formulation here switches the determination of whether the buffer has ever been full to looking for us wrapping the new offset, rather than looking at the old offset.

I think all of these approaches are bogus; if we advance the oldest pointer then we may never actually be "full".  I think the current behaviour only works because the EKF doesn't consume until the buffers have been full at least once.

The supplied tests fail before this PR, pass after (the buffer says it is full on the third element, not the fourth, as @luweiagi pointed out in the issue).

Closes #25316

### CODE REVIEW

File: libraries/AP_NavEKF/EKF_Buffer.cpp [hunk 1]
```diff
@@ -151,13 +152,14 @@ void ekf_imu_buffer::push_youngest_element(const void *element)
         return;
     }
     // push youngest to the buffer
-    _youngest = (_youngest+1) % _size;
+    _youngest = (_youngest+1);
```
Comment (rmackay9) [inline libraries/AP_NavEKF/EKF_Buffer.cpp:155]: I guess this could be _youngest++.

File: libraries/AP_NavEKF/EKF_Buffer.cpp [hunk 2]
```diff
@@ -137,6 +137,7 @@ bool ekf_imu_buffer::init(uint32_t size)
     _size = size;
     _youngest = 0;
     _oldest = 0;
+    _filled = 0;
```
Comment (rmackay9) [inline libraries/AP_NavEKF/EKF_Buffer.cpp:140]: shouldn't this be _filled = false?

**Review by rmackay9 — APPROVED** [review]
LGTM after the two minor fixes

File: libraries/AP_NavEKF/EKF_Buffer.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_NavEKF/EKF_Buffer.cpp]: Fixed

File: libraries/AP_NavEKF/EKF_Buffer.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_NavEKF/EKF_Buffer.cpp]: Fixed

**Comment by luweiagi** [comment]
I think there is a more correct fix at:
https://github.com/ArduPilot/ardupilot/issues/25316#issuecomment-2063060767

**Comment by rmackay9** [comment]
@luweiagi,

Oh oh.  I've added this back for discussion at next week's dev call so that it's not forgotten in case PeterB doesn't have time to look into it before then.

**Pull request merged**